airbyte-server/src/main/java/io/airbyte/server/ConfigDumpImport.java

```diff
@@ -0,0 +1,356 @@
+        LOGGER.info("Setting the DB Airbyte version to : " + targetVersion);
+        postgresPersistence.setVersion(targetVersion);
+
+        // 3. dry run
+        try {
+          checkImport(sourceRoot);
+        } catch (Exception e) {
+          LOGGER.warn("Dry run failed, setting DB version back to initial version");
+          postgresPersistence.setVersion(initialVersion);
+          throw e;
+        }
+
+        // 4. Import Postgres content
+        importDatabaseFromArchive(sourceRoot, targetVersion);
+
+        // 5. Import Configs
+        importConfigsFromArchive(sourceRoot, false);
+        result = new ImportRead().status(StatusEnum.SUCCEEDED);
+      } finally {
+        FileUtils.deleteDirectory(sourceRoot.toFile());
+        FileUtils.deleteQuietly(archive);
+      }
+
+      // identify this instance as the new customer id.
+      TrackingClientSingleton.get().identify();
+      // report that the previous customer id is now superseded by the imported one.
+      previousCustomerIdOptional.ifPresent(
+          previousCustomerId -> TrackingClientSingleton.get().alias(previousCustomerId.toString()));
+    } catch (IOException | JsonValidationException | RuntimeException e) {
+      LOGGER.error("Import failed", e);
+      result = new ImportRead().status(StatusEnum.FAILED).reason(e.getMessage());
+    }
+
+    return result;
+  }
+
+  private void checkImport(Path tempFolder) throws IOException, JsonValidationException {
+    final Path versionFile = tempFolder.resolve(VERSION_FILE_NAME);
+    final String importVersion = Files.readString(versionFile, Charset.defaultCharset())
+        .replace("\n", "").strip();
+    LOGGER.info(String.format("Checking Airbyte Version to import %s", importVersion));
+    if (!AirbyteVersion.isCompatible(targetVersion, importVersion)) {
+      throw new IOException(String
+          .format("Imported VERSION (%s) is incompatible with current Airbyte version (%s).\n" +
+              "Please upgrade your Airbyte Archive, see more at https://docs.airbyte.io/tutorials/upgrading-airbyte\n",
+              importVersion, targetVersion));
+    }
+    checkDBVersion(targetVersion);
+    importConfigsFromArchive(tempFolder, true);
+  }
+
+  // Config
+  private List<String> listDirectories(Path sourceRoot) throws IOException {
+    try (Stream<Path> files = Files.list(sourceRoot.resolve(CONFIG_FOLDER_NAME))) {
+      return files.map(c -> c.getFileName().toString())
+          .collect(Collectors.toList());
+    }
+  }
+
+  private <T> void importConfigsFromArchive(final Path sourceRoot, final boolean dryRun) throws IOException, JsonValidationException {
+    List<String> sourceDefinitionsToMigrate = new ArrayList<>();
+    List<String> destinationDefinitionsToMigrate = new ArrayList<>();
+    final boolean[] sourceProcessed = {false};
+    final boolean[] destinationProcessed = {false};
+    List<String> directories = listDirectories(sourceRoot);
+    Collections.sort(directories);
```

Comment:
i think the purpose of the sort is to make sure the source / destination definitions can processed before the sources / destinations themselves, right? if so can you explain that in a comment?

Reply:
actually i think it's the opposite (even more reason why the comment would help 😆  )

Reply:
to make sure I understand this: if we don't process the X_CONNECTION before the X_DEFINITION, a connector at an older version might be upgrade inadvertently to the latest version without user input?